x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/BaseRunAsSuperuserCommand.java

```diff
@@ -0,0 +1,281 @@
+            throw new UserException(ExitCodes.UNAVAILABLE, "Failed to determine the health of the cluster. ", e);
+        }
+        final int responseStatus = response.getHttpStatus();
+        if (responseStatus != HttpURLConnection.HTTP_OK) {
+            if (responseStatus == HttpURLConnection.HTTP_UNAUTHORIZED && retries > 0) {
+                terminal.println(
+                    Terminal.Verbosity.VERBOSE,
+                    "Unexpected http status [401] while attempting to determine cluster health. Will retry at most "
+                        + retries
+                        + " more times."
+                );
+                Thread.sleep(1000);
+                retries -= 1;
+                checkClusterHealthWithRetries(env, terminal, retries);
+            } else {
+                throw new UserException(
+                    ExitCodes.DATA_ERROR,
+                    "Failed to determine the health of the cluster. Unexpected http status [" + responseStatus + "]"
+                );
+            }
+        } else {
+            final String clusterStatus = Objects.toString(response.getResponseBody().get("status"), "");
+            if (clusterStatus.isEmpty()) {
+                throw new UserException(
+                    ExitCodes.DATA_ERROR,
+                    "Failed to determine the health of the cluster. Cluster health API did not return a status value"
+                );
+            } else if ("red".equalsIgnoreCase(clusterStatus)) {
+                throw new UserException(ExitCodes.UNAVAILABLE, "Cluster health is currently RED.");
+            }
+            // else it is yellow or green so we can continue
+        }
+    }
+
+    HttpResponse.HttpResponseBuilder responseBuilder(InputStream is) throws IOException {
+        final HttpResponse.HttpResponseBuilder httpResponseBuilder = new HttpResponse.HttpResponseBuilder();
+        final String responseBody = Streams.readFully(is).utf8ToString();
+        httpResponseBuilder.withResponseBody(responseBody);
+        return httpResponseBuilder;
+    }
+
+     char[] generatePassword(int passwordLength) {
+        final char[] passwordChars = ("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789~!@#$%^&*-_=+?").toCharArray();
+        char[] characters = new char[passwordLength];
+        for (int i = 0; i < passwordLength; ++i) {
+            characters[i] = passwordChars[secureRandom.nextInt(passwordChars.length)];
+        }
+        return characters;
+    }
+
+    private String generateUsername() {
+        final char[] usernameChars = ("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789").toCharArray();
+        int usernameLength = 8;
+        char[] characters = new char[usernameLength];
+        for (int i = 0; i < usernameLength; ++i) {
+            characters[i] = usernameChars[secureRandom.nextInt(usernameChars.length)];
+        }
+        return "enrollment_autogenerated_" + new String(characters);
+    }
```

Comment:
We generate the two random pieces in multiple places. It would be ideal if we can move this to a place closer to the method that validates these, and reuse that method.

Reply:
That bothered me too a little. I didn't try to move this somewhere because different consumers might have different needs. I.e. we're generating 20 char passwords here but (on purpose) not just alphanumeric chars ( we can iterate on if we agree on this approach too ) but i.e. SetupPasswordsTool generates 20char long alphanumeric passwords. There might be the argument for consistency but one does both temp file realm and reserved user passwords and the other reserved user passwords which doesn't necessarily need the same password strength. 

We can have a method in `org.elasticsearch.xpack.core.security.support.Validation` that generates passwords given a char array and a password length and use that everywhere? But that doesn't move this close to validation as the validation only cares about minimum requirements.

Reply:
My bad.